Added support for Length Visualisation.

aniarya82 opened the pull request:
Using d3's histogram added a plot to display frequency of lengths of sequences in
the hits. Plot also displays relationship between Length of sequences and query
sequence. Simailar to the Gradient Overview Length Visualisation also supports
color gradient based on the E-Values.
Here is an example:
![length-vis3](https://cloud.githubusercontent.com/assets/7463217/16166738/9cfe9aa6-350b-11e6-962e-437d4dab0a3a.png)

yannickwurm commented:
Thanks v much Hiten - this is a step in the right direction - clearly still many shortcomings - please email me when you have a more mature concept

cheers,
Yannick






> On 2016-Jun-17, at 23:48, Hiten Chowdhary <notifications@github.com> wrote:
> 
> Using d3's histogram added a plot to display frequency of lengths of sequences in
> the hits. Plot also displays relationship between Length of sequences and query
> sequence. Simailar to the Gradient Overview Length Visualisation also supports
> color gradient based on the E-Values.
> Here is an example:
> 
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/wurmlab/sequenceserver/pull/281
> 
> Commit Summary
> 
> 	• Added support for Length Visualisation.
> File Changes
> 
> 	• A public/js/length_visualisation.js (174)
> 	• M public/js/report.js (22)
> Patch Links:
> 
> 	• https://github.com/wurmlab/sequenceserver/pull/281.patch
> 	• https://github.com/wurmlab/sequenceserver/pull/281.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

aniarya82 commented:
An updated version
- Fixed the ticks formatting in x-axis.
- Moved legend to bottom of the visualisation.
- Reduced the bins sizes to make graph more precise.
- Changed the format of length-visualasation.js file to make it Object oriented.
![screen shot 2016-06-23 at 2 10 57 pm](https://cloud.githubusercontent.com/assets/7463217/16297159/3fbb21c0-394d-11e6-9723-bf43424e58c1.png)

yeban commented on public/js/report.js:
```diff
@@ -742,6 +743,25 @@ var Hit = React.createClass({
 });
 
 /**
+ * Renders Length Visualisation of all hits per query
```
I guess length visualisation could be of many type - length distribution would a more appropriate name.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,306 @@
+"use strict"
+
+function LengthVisualisation(query, svg_container) {
```
Why does every function and property begin with `_`?

yeban replied:
Why not use ES6 classes instead?

yeban commented on public/js/report.js:
```diff
@@ -742,6 +743,25 @@ var Hit = React.createClass({
+ * Renders Length Visualisation of all hits per query
+ */
+var LengthVis = React.createClass({
+  mixins: [Utils],
```
It's not used here, is it?

yeban commented on public/js/visualisation_helpers.js:
```diff
@@ -0,0 +1,65 @@
+function VisualisationHelpers() {
```
Use [ES6 modules](http://exploringjs.com/es6/ch_modules.html) instead.

yeban commented on public/js/visualisation_helpers.js:
```diff
@@ -0,0 +1,65 @@
+}
+
+/**
+ * Defines how ticks will be formatted.
```
Please could you add an explanation of what this does with 2-3 examples?

yeban commented on public/js/visualisation_helpers.js:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Defines how ticks will be formatted.
+ *
+ * Modified by Priyam based on https://github.com/mbostock/d3/issues/1722.
```
Now that it's out of kablammo folder, better to say - "borrowed from kablammo - modified by ..."

yeban commented on public/js/visualisation_helpers.js:
```diff
@@ -0,0 +1,65 @@
+ *
+ * Modified by Priyam based on https://github.com/mbostock/d3/issues/1722.
+ */
+VisualisationHelpers.prototype._create_formatter  = function (scale, seq_type) {
```
Call it tick_formatter?

yeban commented on public/js/visualisation_helpers.js:
```diff
@@ -0,0 +1,65 @@
+
+// Returns copy of `arr` containing only unique values. Assumes duplicate
+// values always occur consecutively (which they will if `arr` is sorted).
+VisualisationHelpers.prototype._uniq = function (arr) {
```
Please use [`_.uniq`](http://underscorejs.org/#uniq) instead of this helper method.

vivekiitkgp commented:
@aniarya82 Let me know if I am missing something. Since the visualization above is a histogram, the function is essentially binning the data and showing the distribution. Which means that by controlling the number of bins to use, you can make the distribution smooth or rough. Did you experiment with this? What's the currently used strategy?

vivekiitkgp commented:
@aniarya82 Also, I think it'd be handy to have a hover tooltip for histogram bars that shows the number of HSPs, Identity and Query coverage values.

yannickwurm commented:
- [ ] start X axis with 0, end it with 1.05*max(hits, query)  (to have extra space on either side).
 - [ ] every hit's must have a different color (right now a bit has the color of the minimum evalue  - this can be misleading
 - [ ] use less vertical space
 - [ ] query should be more visible
 - [ ] too many lablels. Perhaps can have fewer ticks:use (perhaps) between 4 labels and max 12 ticks in total? (like kablammo). And/or not every tick needs a label.
 - [ ] keep 1.0 rather than 1 (if also showing e.g. 0.8 or 0.5)
 - [ ] add bp and aa when appropriate. 
 - [ ] Consider whether we should shift bins by 50% of tick width - like [here](https://learnandteachstatistics.files.wordpress.com/2012/11/histogram_0.png)
 - [ ] no need for a separate legend. 
 - [ ] beginning and end of x axis should be vertically aligned with first and last ticks of overview graph. 
 (no need for funky legend colors because we have a single continuous variable with no biologically meaningful transitionpoints).

aniarya82 commented:
Changes in length distribution chart:
- Reduced the vertical space taken up by chart
- changes x-axis domain ([0,1.05*max(hits,query)])
- altered the color of the query line to make it more legible
- reduced the number of ticks labelled
- added 'aa' and 'bp' when appropriate
- removed the legend
![length-vis](https://cloud.githubusercontent.com/assets/7463217/16479053/6089043a-3ebb-11e6-8235-b86a12e84fbc.png)

yannickwurm commented:
Glad to see this is possilble :)
Can you tick the checkboxes in the summary of our discussion above? 
Also:
- [ ] flip the colors so the strongest is at the bottom
- [ ] no "k" abbreviation for amino acids

Cheers!
y

yeban commented:
This looks great @aniarya82! We shouldn't include 'aa' / 'bp' for 0. Could you also try to show tooltips indicating hit name, length and evalue? Can we have the same tooltip generation code reused in both graphical overview and length visualisation?

@yannickwurm Proteins can also get mighty big (e.g., actin and all). The tick formatter used here is adaptive: it's when the numbers get big that they are reformatted with SI prefix - 'k', 'm', 'g', for 10^3, 10^6, and 10^9. We inherited the technique from Kablammo and use it in graphical overview too (one of the other things in master). Instead of removing 'k' for 'aa', how about rendering '2.5k aa' instead of '2.5 kaa'. Or, we don't do 'k' SI prefix till 50,000 for both aa and bp (knowing that the largest protein we know of is 33,000 aa long).

aniarya82 commented:
Few updates:
- Flipped colours, now strongest hit is at bottom
- Removed 'k' abbreviation for amino acids
- Removed 'aa' / 'bp' for 0
- Added tooltips indication hit name, evalue and length.
![length-vis 2](https://cloud.githubusercontent.com/assets/7463217/16485714/0d416d80-3ede-11e6-8730-1f91a501ebb4.png)

yeban commented:
I just tried this locally. The y-axis looks like in one of the first pics you posted:

<img width="500" alt="screen shot 2016-06-30 at 4 35 28 pm" src="https://cloud.githubusercontent.com/assets/90373/16486215/ce4b90d0-3ee0-11e6-97a1-654c64190ca6.png">

yeban commented:
I think the tooltips should be displayed without a delay.

aniarya82 commented:
Minor updates:
- Y-axis formatting fixed.
- reduced tooltips delay to 0.
![screen shot 2016-06-30 at 8 07 17 pm](https://cloud.githubusercontent.com/assets/7463217/16492175/7e132fec-3efe-11e6-94ef-eb371bc467ff.png)

yeban commented:
- [ ] It is not very meaningful to have ticks on Y axis apart from 1, 2, 3, N.
- [ ] I think we should show only one 0 for both axes combined.
- [ ] Can you align it with graphical overview?

yeban commented on public/js/visualisation_helpers.js:
```diff
@@ -0,0 +1,57 @@
+import _ from 'underscore';
+
+export default class VisualisationHelpers {
```
We don't want a class here. Just a collection of methods in a file. Import in graph.js or length_visualisation.js as `import * as Helpers from './visualisation_helpers.js'` and use as `Helpers.tick_formatter`.

yeban commented on public/js/report.js:
```diff
@@ -742,6 +745,24 @@ var Hit = React.createClass({
+  },
+  render: function () {
+    return(
+      <div ref="plot_test">
```
I would prefer the name "svgContainer". Also, give it a class: length-distribution.

yeban commented on public/js/report.js:
```diff
@@ -742,6 +745,24 @@ var Hit = React.createClass({
+    return $(React.findDOMNode(this.refs.plot_test));
+  },
+  componentDidMount: function () {
+    this.plot_test = new LengthVisualisation(this.props.query, this.plot_svg(), this.props.algorithm);
```
Store the object in `this.graph`.

yeban commented on public/js/report.js:
```diff
@@ -742,6 +745,24 @@ var Hit = React.createClass({
+ * Renders Length Distribution of all hits per query
+ */
+var LengthVis = React.createClass({
+  plot_svg: function () {
```
Rename it to `svgContainer` in line with GraphicalOverview and Kablammo components.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+    this.visualisation_helpers = new VisualisationHelpers();
+
+    this._margin = {top: 30, right: 60, bottom: 70, left: 25};
+    this._width = 838 - this._margin.left - this._margin.right;
```
Shouldn't it be svg_container.width() instead of 838 here?

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+  constructor(query, svg_container, algorithm) {
+    this._svg_container_d3 =  d3.select(svg_container[0]);
+    this._svg = {};
+    this._query = query;
```
`_` is used to indicate that this property is private. We don't need to do that for query.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+    this._svg = {};
+    this._query = query;
+    this._seq_type = this.get_seq_type(algorithm);
+    this._hits = query.hits;
```
Redundant - you can always do `this.query.hits`.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+    return data;
+  }
+
+  get_seq_type(algorithm) {
```
Shouldn't this be moved to helpers?

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+    this.draw_visualisation();
+  }
+
+  fill_data() {
```
This is not a meaningful name. Should be `hit_lengths`.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  fill_data() {
+    var data = [];
```
`map` returns an array. You are discarding `map`'s return value and using it like `each`.

`return _.map(this.query, _.iteratee('length'))`

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+
+  draw_rectangles() {
+    var self = this;
+    var getEvalue = function(bin) {
```
Unused function it seems?

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+    });
+  }
+
+  get_evalue(bin) {
```
Unused function?

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  create_axis() {
+    var formatter = this.tick_formatter(this._seq_type.subject_seq_type);
```
Why are we not using `tick_formatter` from visualisation helpers?

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+    return returnString;
+  }
+
+  fill_data2() {
```
This function is very unclear - including names like data2, reusing the same variable name in different contexts, a method called set_evalue that matches based on length.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+        .attr('transform','rotate(-45)');
+  }
+
+  create_axis() {
```
Creates both x and y axis. Should be more appropriately named - create_axes.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+        .text('Number of Sequences');
+  }
+
+  draw_legend() {
```
Is this the exact same code from graphical overview? If so it's safe to remove it I guess since we are not drawing a legend here.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+import VisualisationHelpers from './visualisation_helpers';
+
+export default class LengthVisualisation {
+  constructor(query, svg_container, algorithm) {
```
Let's call it `$svg_container` to be clear that it's jquery wrapped DOM node.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+export default class LengthVisualisation {
+  constructor(query, svg_container, algorithm) {
+    this._svg_container_d3 =  d3.select(svg_container[0]);
+    this._svg = {};
```
`this._svg` is an object with only one property (`d3` - L36). And `this._svg.d3` is same as `this._svg_container_d3`. Right? Can we have just one `this.svg` instead?

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,339 @@
+        .range([0, this._width]);
+    this._bins = this.create_bins();
+    this._data2 = this.fill_data2();
+    this._y_scale = d3.scale.linear()
```
`this._y_scale` is used only once in `create_axis`. Why not define the variable in `create_axis` method?

yannickwurm commented:
- [ ] labels are not fully centered under tickmarks
 - [ ] unlabelled ticks should be lighter or thinner or shorter than those with labels 
 - [ ] y axis label should be vertically centered with regards to Y axis. Actually no: remove axis labels and instead use an informative overall title for the graph. This will allow the width of the x axis to be identical to the graph above. 
 - [ ] a rounding error (or similar) seems to introduce white lines between hit boxes in some cases. And to make some ticks thicker than others. Perhaps activating antialiasing would resolve this?
 - [ ] in old overview visualisation, try putting 0px or 1px between hits so as to vertically compress results.
 - [ ] vertically take up only 30% of current space
 - [ ] check what standard practice is in other beuatiful histogram and bar charts regarding where the axes end. If apporpriate, do what they do (e.g. extend the y and x axis to the next labelled tick).

yannickwurm commented:
-[ ] if there is a good code checker (for syntax, accuracy & svg integrity) please use it.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,247 @@
+  update_data() {
+    var self = this;
+    var data2 = [];
+    var set_index = function(length) {
```
There's probably a function in underscorejs to do this.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,247 @@
+    this._seq_type = Helpers.get_seq_type(algorithm);
+
+    this._margin = {top: 30, right: 60, bottom: 70, left: 25};
+    this._width = $svg_container.width() - this._margin.left - this._margin.right;
```
`width` and `height` are used only inside the constructor. They thus don't need to be a property.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  draw_scales() {
+    var self = this;
```
Doesn't look like assigning `this` to `self` is needed here.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  create_labels() {
+    var self = this;
```
`self = this` not needed here either, I think.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,247 @@
+    return bins;
+  }
+
+  prettify_evalue(evalue) {
```
This functionality is needed in at least 4 places that I can remember: tabular summary, alignment overview (tooltip), stats table before pairwise alignment. And now length distribution. Could you extract this into a helper method and use the same throughout. Treat `report.js`' `prettify_float` to be authoritative.

Similarly, tooltip generation should be shared between this and alignment overview.

yeban commented on public/js/length_visualisation.js:
```diff
@@ -0,0 +1,247 @@
+        .domain([0, d3.max(this._bins, function(d) { return d.length })])
+        .range([this._height, 0]);
+    this._axis = this.create_axes();
+    this.svg = this._svg_container_d3.insert('svg', ':first-child')
```
Doesn't the right hand side expression return `this._svg_container_d3`?

yeban commented:
Great work! Thanks.

Could you add anything left to take care of to #273 along with other stuff that we discussed (redraw on resize, svg & png download buttons)?
